tests/csv_datums/test_csv_sources.py

```diff
@@ -90,35 +124,198 @@ def source_interval():
+        res['result']['last'] = str(last)
+        return res
+
+    return factory
+
+
+@pytest.fixture
+def randomness(monkeypatch):
+    s = RandomnessSpy()
+    monkeypatch.setattr(module_under_test, 'random', s)
+    return s
 
 
-def test_kraken_source_returns_validated_and_adapted_data(source, source_interval, requests):
-    requests.set_get_response(json={'some': 'data'})
-    adapted = Adapted(json={'some': 'data'}, with_interval=source_interval)
-    assert source.query(since=1559347200) == Validated(csv_datums_from(adapted))
+def make_get(url, **params):
+    return GetRequest(url, params)
 
 
 def csv_datums_from(adapted):
     return CsvDatums(adapted.interval, adapted)
+
+
+def expand_to_pair(*values):
+    return [[v, v, 1559347200, v, v, v] for v in values]
+
+
+@pytest.mark.usefixtures("server_time")
+class TestKrakenSource:
+    def test_kraken_source_builds_correct_url(self, source, requests):
+        source.query(since=1559347200)
+        assert requests.received_get == make_get("https://api.kraken.com/0/public/Trades", pair="xbtusd",
+                                                 since=1559347200000000000)
+
+    @pytest.mark.parametrize("invalid", [dict(), dict(result={'pair': []}), dict(error=[])])
+    def test_invalid_kraken_api_format(self, requests, source, invalid):
+        requests.set_get_response(json=invalid)
+        with pytest.raises(InvalidFormatError):
+            source.query(since=1559347200)
+
+    def test_error_in_response(self, requests, source, make_json):
+        requests.set_get_response(json=make_json({'pair': [['data']]}, with_error=['some error']))
+        with pytest.raises(ResponseError):
+            source.query(since=1559347200)
+
+    def test_kraken_source_returns_validated_and_adapted_data(self, source, source_interval, requests, validation,
+                                                              make_json):
+        requests.set_get_response(json=make_json({'pair': expand_to_pair(1)}))
+        adapted = AdaptedData(trades=[[1, 1, 1559347200]], with_interval=source_interval)
+        assert source.query(since=1559347200) == csv_datums_from(adapted)
+        assert validation.data == csv_datums_from(adapted)
+
+    def test_kraken_source_queries_data_up_to_last_incomplete_interval(self, source, source_interval, requests,
+                                                                       server_time, make_json):
+        server_time.set_current_time(1559347200 + source_interval * 3 * 60 + 10)
+        requests.set_get_responses(
+            make_json({'pair': expand_to_pair(1)}, last=to_nano_sec(1559347200 + source_interval * 60)),
+            make_json({'pair': expand_to_pair(2)}, last=to_nano_sec(1559347200 + source_interval * 1.5 * 60)),
+            make_json({'pair': expand_to_pair(3)}, last=to_nano_sec(1559347200 + source_interval * 3 * 60)),
+            make_json({'pair': expand_to_pair(4)}, last=to_nano_sec(1559347200 + source_interval * 3 * 60 + 1)),
```

Comment:
here it would be possible to use a proper from to range in the `expand_to_trades` function